### DISCUSSION TITLE
Apple M1 PyQtGraph Performance

### DISCUSSION THREAD

**Discussion by j9ac9k** [pyqtgraph/pyqtgraph#2044] [General]
Tagging @Carreau

@Carreau offered to run some PyQtGraph benchmarks on his new Apple M1 Pro chip.  I've been curious how numpy/Qt will behave on there. Supposedly Qt 6.2 supports the macOS Universal binaries, so we _should_ get "native" performance that way.  I'm actually not sure what the status of numpy is there; regardless, figured I would open up this issue to test things out.

You'll need a virtual environment with `numpy`, `scipy` and either PyQt6 6.2 or PySide6 6.2 (latest versions of either one).  I would recommend running the benchmarks from a clone of the pyqtgraph repo, installed in editable mode while on the current commit of the master branch.

There are 4 benchmarks can be run as follows:

```bash
python examples/PlotSpeedTest.py
python examples/ScatterPlotSpeedTest.py
python examples/VideoSpeedTest.py  # someone else may have some suggested arguments to pass along here
python examples/PColorMeshItem.py
```

The FPS might jump around a bit, and if the mouse remains steady w/ no keyboard entry, after a few seconds you may notice the FPS values going up.

Anyway if you would be willing to post screenshots (or approximate FPS numbers) I would be super curious, thanks for offering to indulge me on this!

**Comment by Carreau** [comment]
Sorry what I'll ask might be dumb questions, but should I install qt from https://www.qt.io/download-thank-you and pip install the  rest ? Or is there a conda package?

It's weird as the DMG is called "qt-unified-macOS-x64-4.1.1-online.dmg" I would have expected a -6.2- somewhere in the name

**Reply by j9ac9k** [reply]
oh heavens no, just via pip

```
pip install pyqt6
```

EDIT: sorry should have elaborated on that, yeah navigating the Qt website for what tool to get...that's a head ache you want no part of, if you were to install the Qt framework (not through python bindings) the easiest way would be via homebrew, but since we're not looking to build anything, we don't need to install it.

**Reply by Carreau** [reply]
<details>
<pre>
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro ~ % conda install python pip
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

The following NEW packages will be INSTALLED:
  libzlib            conda-forge/osx-arm64::libzlib-1.2.11-hee7b306_1013
  ncurses            conda-forge/osx-arm64::ncurses-6.2-h9aa5885_4
  openssl            conda-forge/osx-arm64::openssl-3.0.0-h3422bc3_1
  pip                conda-forge/noarch::pip-21.3.1-pyhd8ed1ab_0
  readline           conda-forge/osx-arm64::readline-8.1-hedafd6a_0
  setuptools         conda-forge/osx-arm64::setuptools-58.2.0-py39h2804cbe_0
  tzdata             conda-forge/noarch::tzdata-2021e-he74cb21_0
  wheel              conda-forge/noarch::wheel-0.37.0-pyhd8ed1ab_1
  xz                 conda-forge/osx-arm64::xz-5.2.5-h642e427_1
  zlib               conda-forge/osx-arm64::zlib-1.2.11-hee7b306_1013


Preparing transaction: done
Verifying transaction: done
Executing transaction: done
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro ~ % pip install pyqt6
Collecting pyqt6
  Downloading PyQt6-6.2.0-cp36-abi3-macosx_10_14_universal2.whl (11.5 MB)
     |████████████████████████████████| 11.5 MB 9.0 MB/s
     |████████████████████████████████| 30.5 MB 21.0 MB/s
Collecting PyQt6-sip<14,>=13.1
  Downloading PyQt6_sip-13.1.0.tar.gz (89 kB)
     |████████████████████████████████| 89 kB 17.2 MB/s
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: PyQt6-sip
  Building wheel for PyQt6-sip (pyproject.toml) ... done
  Created wheel for PyQt6-sip: filename=PyQt6_sip-13.1.0-cp39-cp39-macosx_11_0_arm64.whl size=45761 sha256=d9afb1f7bed0bc074a2858c8937909ac79b7f5ed9ad9f3be49895eeb50b64c6b
  Stored in directory: /Users/bussonniermatthias/Library/Caches/pip/wheels/ed/5b/e7/5faedd9396173479140caf56e901ab96bbd82b3ff016668ac3
Successfully built PyQt6-sip
Installing collected packages: PyQt6-sip, PyQt6-Qt6, pyqt6
Successfully installed PyQt6-Qt6-6.2.0 PyQt6-sip-13.1.0 pyqt6-6.2.0
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro ~ % cd pyqtgr
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro ~ % cd
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro ~ % cd dev
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro dev % cd pyqtgraph
(pyqt6) bussonniermatthias@Matthiass-MacBook-Pro pyqtgraph % pip install -e .
Obtaining file:///Users/bussonniermatthias/dev/pyqtgraph
  Preparing metadata (setup.py) ... done
Collecting numpy>=1.17.0
  Downloading numpy-1.21.3-cp39-cp39-macosx_11_0_arm64.whl (12.4 MB)
     |████████████████████████████████| 12.4 MB 9.0 MB/s
Installing collected packages: numpy, pyqtgraph
  Running setup.py develop for pyqtgraph
Successfully installed numpy-1.21.3 pyqtgraph-0.12.3
</pre>
</details>

<img width="752" alt="Screen Shot 2021-10-26 at 11 08 26 PM" src="https://user-images.githubusercontent.com/335567/139009501-b6620cba-d499-4410-8222-8901f9d61f79.png">

<img width="1042" alt="Screen Shot 2021-10-26 at 11 06 53 PM" src="https://user-images.githubusercontent.com/335567/139009495-7b027755-d710-41dd-89eb-0a4dbd638418.png">
<img width="1049" alt="Screen Shot 2021-10-26 at 11 08 04 PM" src="https://user-images.githubusercontent.com/335567/139009500-cff08107-d383-4c1d-a7ee-78950ea8e453.png">

<img width="807" alt="Screen Shot 2021-10-26 at 11 11 31 PM" src="https://user-images.githubusercontent.com/335567/139009592-563de924-1e75-450d-bb14-a324ee4aa75a.png">

Is that of any help ? I can try with intel emulation tomorrow.

**Reply by j9ac9k** [reply]
Thanks so much @Carreau  ...we're looking at a rather amazing shift in performance, I'm looking at 2x performance boost (vs my 2019  2 GHz 4 core CPU) for everything except the `PColorMeshItem`; and if I remember your twitter message right, you got the _base_ model CPU, that's amazing.

 @outofculture @campagnola  figured you all might be interested in these results for `VideoSpeedTest.py` given how much that impacts your work.
 
 @danielhrisca @lidstrom83 check out those ScatterPlotSpeedTest numbers 😍

**Reply by danielhrisca** [reply]
Was this in CPU mode or using OpenGL? 

Edit: I'm getting about 480FPS on a Core(TM) i7-8850H CPU (45W) so the M1 is a good step-up

**Reply by Carreau** [reply]
Yes, base model, M1 Pro 16G. 

> Was this in CPU mode or using OpenGL?

I have no clue. I just installed and ran the commands @j9ac9k asked. Let me know if I should do any tweaks.

**Reply by j9ac9k** [reply]
OpenGL was not used here, I don't think we have demonstrated OpenGL to be particularly more performant anywhere (likely because of our implementation) ....exception for thick line rendering, but that might not be true any more.

I am curious what OpenGL performance is like on the M1 variants, but I don't have a good test here (OpenGL expertise is something very much lacking between the active maintainers right now, if you're interested in contributing to pyqtgraph and have experience with OpenGL, please message me haha.

**Comment by psobolewskiPhD** [comment]
For comparison here's 2020 M1 MBPro (16Gb Ram).
(numpy, scipy via mamba, pyqt6, pyqtgraph via pip)
<img width="1042" alt="image" src="https://user-images.githubusercontent.com/76622105/139310990-d921e2ce-afae-4383-b526-239952729031.png">
<img width="1049" alt="image" src="https://user-images.githubusercontent.com/76622105/139311112-d7229457-09d6-4aa7-9844-0742055ec9b2.png">
<img width="807" alt="image" src="https://user-images.githubusercontent.com/76622105/139311298-f7d696f7-3a12-4fbc-98dd-bac4a614e2d1.png">
The mesh runs fine, but I'm not seeing a FPS indicator.
BTW, i see just 1 python process, ~100% CPU.
So I think this largely confirms that the only difference between the M1 and M1Pro is in the number of performance cores and GPU cores. But the Firestorm cores themselves are essentially the same, so this single thread bench runs the same.
Need some GPU benches! Or some benches that test the unified memory architecture...

**Reply by j9ac9k** [reply]
Hi @psobolewskiPhD  Thanks for posting these!

The PColorMeshItem.py was only recently changed to include benchmark info, it might have been done after the 0.12.3 release.  That change also included a substantial performance improvement with it; and it will be part of the next release.

If you want to test OpenGL related performance, you can run

```
python examples/PlotSpeedTest.py --opengl
```

This will use the OpenGL variant of PlotCurveItem.  Fair warning, our implementation likely could use a lot of work, and it's not as feature-complete as the non-opengl variant.

> Need some GPU benches! Or some benches that test the unified memory architecture...

Do you have some idea if there is something within the python scientific computing ecosystem that would potentially stress this architecture?  PyQtGraph does have some optional CuPy support, but as we can't have CUDA on macOS that's not particularly helpful.

We do use numba in some-places with respect to the VideoSpeedTest; do you know if numba would benefit there?

**Reply by psobolewskiPhD** [reply]
I tried, that—warnings galore:
```
 --- exception caught here ---
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/debug.py", line 110, in w
    func(*args, **kwds)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/graphicsItems/PlotCurveItem.py", line 508, in paint
    self.paintGL(p, opt, widget)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/graphicsItems/PlotCurveItem.py", line 568, in paintGL
    import OpenGL.GL as gl
ModuleNotFoundError: No module named 'OpenGL'
  printExc('Ignored exception:')
/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/debug.py:112: RuntimeWarning: Ignored exception:
Traceback (most recent call last):
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/examples/PlotSpeedTest.py", line 206, in <module>
    pg.exec()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/Qt/__init__.py", line 409, in exec_
    return app.exec() if hasattr(app, 'exec') else app.exec_()
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/examples/PlotSpeedTest.py", line 189, in update
    app.processEvents(QtCore.QEventLoop.ProcessEventsFlag.AllEvents)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/widgets/GraphicsView.py", line 143, in paintEvent
    return super().paintEvent(ev)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/examples/PlotSpeedTest.py", line 81, in paint
    return super().paint(painter, opt, widget)
  File "/Users/piotrsobolewski/Dev/miniforge3/envs/pyqtgraph-test/lib/python3.9/site-packages/pyqtgraph/debug.py", line 112, in w
    printExc('Ignored exception:')
```
And much lower performance:
<img width="1042" alt="image" src="https://user-images.githubusercontent.com/76622105/139319833-5580d321-b4e4-4830-80c3-8deb75001252.png">

Is there something else to install?

**Reply by j9ac9k** [reply]
oh, yes there is 😆 

```
pip install pyopengl
```

Performance on the lines will be lower with the openGL enabled and otherwise default configs (this is likely due to our use of opengl, this is an area us maintainers collectively have little experience in, if someone with opengl knowledge comes across this, and wants to help, please reach out!)

I'm more interested in what the performance difference is between the ARM and Intel CPUs here.  I can provide equivalent results for my intel 2019 model macbook pro.

**Comment by psobolewskiPhD** [comment]
Derp! Sorry!
Here's the output:
<img width="1042" alt="image" src="https://user-images.githubusercontent.com/76622105/139322217-7bce1e85-f7d8-4896-a96d-f220d4e36e06.png">

**Reply by j9ac9k** [reply]
Here's my result on the intel system:

![image](https://user-images.githubusercontent.com/646398/139322429-36ea916f-10c6-443c-84e9-10e02172c046.png)

not sure if the boost is from numpy end or the rendering end... but suffice to say that OpenGL renderer seems to work and not take major performance hits 👍🏻

**Comment by psobolewskiPhD** [comment]
And here's with extra
`pip install PyOpenGL-accelerate`

<img width="1042" alt="image" src="https://user-images.githubusercontent.com/76622105/139322318-5078a845-1ca0-4990-ad81-aee3ece12340.png">
